shared/triage_metadata.go

```diff
@@ -0,0 +1,257 @@
+// Copyright 2019 The WPT Dashboard Project. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+//go:generate mockgen -destination sharedtest/triage_metadata_mock.go -package sharedtest github.com/web-platform-tests/wpt.fyi/shared TriageMetadataInterface
+
+package shared
+
+import (
+	"context"
+	"errors"
+	"math/rand"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/google/go-github/v28/github"
+	"gopkg.in/yaml.v2"
+)
+
+// TriageMetadataInterface encapsulates triage() method for testing purpose.
+type TriageMetadataInterface interface {
+	Triage(metadata MetadataResults) error
+}
+
+type triageMetadata struct {
+	ctx context.Context
+	metadataGithub
+	logger     Logger
+	httpClient *http.Client
+}
+
+type metadataGithub struct {
+	githubClient *github.Client
+	authorName   string
+	authorEmail  string
+	wptmetadataGitHubInfo
+}
+
+type wptmetadataGitHubInfo struct {
+	sourceOwner   string
+	sourceRepo    string
+	commitMessage string
+	commitBranch  string
+	baseBranch    string
+	prRepoOwner   string
+	prRepo        string
+	prBranch      string
+	prSubject     string
+	prDescription string
+}
+
+func getNewCommitBranchName(ctx context.Context, client *github.Client, sourceOwner string, sourceRepo string) string {
+	commitBranch := "auto-triage-branch" + generateRandomInt()
+	// If the commitBranch name already exist, genereate a new one. We consider that an error means the branch has not been found and needs to
+	// be created.
+	for ref, _, err := client.Git.GetRef(ctx, sourceOwner, sourceRepo, "refs/heads/"+commitBranch); err == nil && ref != nil; {
```

Comment:
Note that this could spin forever if we are getting real errors back from `client.Git.GetRef` (what if we hit an API limit?). I would suggest defensively only trying this loop a few times and bailing if we hit the limit. Real conflicts should be pretty rare (10,000 numbers to choose uniformly from, and in theory only a handful of PR branches exist at any time).

Reply:
By looking at the [client.Git.GetRef method](https://github.com/google/go-github/blob/master/github/git_refs.go#L59), I suspect that there will never be a case where an error occurs but (err == nil && ref != nil) is true. But it certainly doesn't hurt to code defensively here. 

Since the GitHub API rate limit is 5000/hour and it is rare to hit a real conflict, I will only allow this method to loop 10